compiler/rustc_builtin_macros/src/assert.rs

```diff
@@ -42,27 +44,37 @@ pub fn expand_assert<'cx>(
         } else {
             // Before edition 2021, we call `panic!()` unqualified,
             // such that it calls either `std::panic!()` or `core::panic!()`.
-            Path::from_ident(Ident::new(sym::panic, sp))
-        };
-        // Pass the custom message to panic!().
-        cx.expr(
-            sp,
+            Path::from_ident(Ident::new(sym::panic, call_site_span))
+        }
+    };
+
+    // Simply uses the user provided message instead of generating custom outputs
+    let expr = if let Some(tokens) = custom_message {
+        let then = cx.expr(
+            call_site_span,
             ExprKind::MacCall(MacCall {
-                path,
+                path: panic_path(),
                 args: P(MacArgs::Delimited(
-                    DelimSpan::from_single(sp),
+                    DelimSpan::from_single(call_site_span),
                     MacDelimiter::Parenthesis,
                     tokens,
                 )),
                 prior_type_ascription: None,
             }),
-        )
-    } else {
+        );
+        cx.expr_if_not(call_site_span, cond_expr, then, None)
+    }
+    // If `generic_assert` is enabled, generates rich captured outputs
+    else if let Some(features) = cx.ecfg.features && features.generic_assert {
```

Comment:
This is a bad practice, `#![feature]` attributes are supposed to never or almost never change behavior, only to report or not report stability errors.

During stabilization the behavior will switch anyway, and not switching this earlier will only result in the feature having not enough coverage.

It should only be gated like this if it's not fully implemented and has known issues, but then it should have a FIXME explaining it.

Reply:
I'd suggest running this PR through crater (or at least through rustc bootstrap) with the new assert behavior enabled to see whether it will fail somewhere.

Reply:
There are three things that still need to be implemented

1. Constant environments
2. Avoid copying structures when methods don't take ownership. For example, `PartialEq`.
3. Prevent large copies when a method takes ownership. For example, `[u128; 4096]` -> https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b68e519965f9371b1dff653454731811